cmd/kubernetes-discovery/pkg/apiserver/apiserver.go

```diff
@@ -67,26 +92,108 @@ func (c *Config) SkipComplete() completedConfig {
 func (c completedConfig) New() (*APIDiscoveryServer, error) {
+	informerFactory := informers.NewSharedInformerFactory(
+		internalclientset.NewForConfigOrDie(c.Config.GenericConfig.LoopbackClientConfig),
+		clientset.NewForConfigOrDie(c.Config.GenericConfig.LoopbackClientConfig),
+		5*time.Minute, // this is effectively used as a refresh interval right now.  Might want to do something nicer later on.
+	)
+
+	// most API servers don't need to do this, but we need a custom handler chain to handle the special /apis handling here
+	c.Config.GenericConfig.BuildHandlerChainsFunc = (&handlerChainConfig{
+		informers: informerFactory,
+	}).handlerChain
+
 	genericServer, err := c.Config.GenericConfig.SkipComplete().New() // completion is done in Complete, no need for a second time
 	if err != nil {
 		return nil, err
 	}
 
 	s := &APIDiscoveryServer{
-		GenericAPIServer: genericServer,
+		GenericAPIServer:   genericServer,
+		handledAPIServices: sets.String{},
+		lister:             informerFactory.Apiregistration().InternalVersion().APIServices().Lister(),
 	}
 
 	apiGroupInfo := genericapiserver.NewDefaultAPIGroupInfo(apiregistration.GroupName)
 	apiGroupInfo.GroupMeta.GroupVersion = v1alpha1.SchemeGroupVersion
-
 	v1alpha1storage := map[string]rest.Storage{}
 	v1alpha1storage["apiservices"] = apiservicestorage.NewREST(c.RESTOptionsGetter.NewFor(apiregistration.Resource("apiservices")))
-
 	apiGroupInfo.VersionedResourcesStorageMap["v1alpha1"] = v1alpha1storage
 
 	if err := s.GenericAPIServer.InstallAPIGroup(&apiGroupInfo); err != nil {
 		return nil, err
 	}
 
+	apiserviceRegistrationController := NewAPIServiceRegistrationController(informerFactory.Apiregistration().InternalVersion().APIServices(), s)
+
+	s.GenericAPIServer.AddPostStartHook("start-informers", func(context genericapiserver.PostStartHookContext) error {
+		informerFactory.Start(wait.NeverStop)
+		return nil
+	})
+	s.GenericAPIServer.AddPostStartHook("apiservice-registration-controller", func(context genericapiserver.PostStartHookContext) error {
+		apiserviceRegistrationController.Run(wait.NeverStop)
+		return nil
+	})
+
 	return s, nil
 }
+
+// handlerChainConfig is the config used to build the custom handler chain for this api server
+type handlerChainConfig struct {
+	informers informers.SharedInformerFactory
+}
+
+// handlerChain is a method to build the handler chain for this API server.  We need a custom handler chain so that we
+// can have custom handling for `/apis`, since we're hosting discovery differently from anyone else and we're hosting
+// the endpoints differently, since we're proxying all groups except for apiregistration.k8s.io.
+func (h *handlerChainConfig) handlerChain(apiHandler http.Handler, c *genericapiserver.Config) (secure, insecure http.Handler) {
+	// add this as a filter so that we never collide with "already registered" failures on `/apis`
+	handler := WithAPIs(apiHandler, h.informers.Apiregistration().InternalVersion().APIServices())
```

Comment:
Why not call the default chain builder with your wrapper `apiHandler`? Don't particularly like the code duplication below.